src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef STAN_ANALYZE_MCMC_COMPUTE_POTENTIAL_SCALE_REDUCTION_HPP
+#define STAN_ANALYZE_MCMC_COMPUTE_POTENTIAL_SCALE_REDUCTION_HPP
+
+#include <stan/math/prim/mat.hpp>
+#include <stan/analyze/mcmc/autocovariance.hpp>
+#include <stan/analyze/mcmc/split_chains.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <algorithm>
+#include <cmath>
+#include <vector>
+#include <limits>
+
+namespace stan {
+namespace analyze {
+
+  /**
+   * Computes the potential scale reduction (Rhat) for the specified
+   * parameter across all kept samples.
+   *
+   * See more details in Stan reference manual section "Potential
+   * Scale Reduction". http://mc-stan.org/users/documentation
+   *
+   * Current implementation assumes chains are all of equal size and
+   * draws are stored in contiguous blocks of memory.
+   *
+   * @param draws stores pointers to arrays of chains
+   * @param sizes stores sizes of chains
+   * @return potential scale reduction for the specified parameter
+   */
+  inline
+  double compute_potential_scale_reduction(std::vector<const double*> draws,
+                                           std::vector<size_t> sizes) {
+    int num_chains = sizes.size();
+    size_t num_draws = sizes[0];
+    for ( int chain = 1; chain < num_chains; ++chain ) {
+      num_draws = std::min(num_draws, sizes[chain]);
+    }
+
+    Eigen::VectorXd draw_val = Eigen::VectorXd::Random(num_chains);
+    for ( int chain = 0; chain < num_chains; chain++ ) {
+      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
+        draw(draws[chain], sizes[chain]);
+
+      for ( int n = 0; n < num_draws; n++ ) {
+        if ( !boost::math::isfinite(draw(n)) ) {
+          return std::numeric_limits<double>::quiet_NaN();
+        }
+      }
+
+      if ( draw.isApproxToConstant(draw(0)) ) {
+        draw_val(chain) = draw(0);
+      }
+    }
+
+    if ( draw_val.isApproxToConstant(draw_val(0)) ) {
```

Comment:
Same comment as above -- if everything is constant then do we want a NaN here instead of a "perfect" value?

Reply:
Same as above.  Should we over rule the consistency argument in favor of returning NaN here?

Reply:
Let me start a Discourse thread appealing for opinion on this.  Worst case I'll approve for the consistency of the current behavior and we'll fix with a separate PR.